pkg/scrapers/twitter/tweets.go

```diff
@@ -3,212 +3,184 @@ package twitter
 func ScrapeTweetsByQuery(query string, count int) ([]*TweetResult, error) {
-	scraper := auth()
-	var tweets []*TweetResult
-	var lastError error
+	once.Do(initializeAccountManager)
 
-	if scraper == nil {
-		return nil, fmt.Errorf("there was an error authenticating with your Twitter credentials")
+	getAuthenticatedScraper := func() (*twitterscraper.Scraper, *TwitterAccount, error) {
+		account := accountManager.GetNextAccount()
+		if account == nil {
+			return nil, nil, fmt.Errorf("all accounts are rate-limited")
+		}
+		scraper := Auth(account)
+		if scraper == nil {
+			logrus.Errorf("authentication failed for %s", account.Username)
+			return nil, account, fmt.Errorf("authentication failed for %s", account.Username)
+		}
+		return scraper, account, nil
 	}
 
-	// Set search mode
-	scraper.SetSearchMode(twitterscraper.SearchLatest)
-
-	// Perform the search with the specified query and count
-	for tweetResult := range scraper.SearchTweets(context.Background(), query, count) {
-		if tweetResult.Error != nil {
-			lastError = tweetResult.Error
-			logrus.Warnf("[+] Error encountered while scraping tweet: %v", tweetResult.Error)
-			if strings.Contains(tweetResult.Error.Error(), "Rate limit exceeded") {
-				return nil, fmt.Errorf("Twitter API rate limit exceeded (429 error)")
+	scrapeTweets := func(scraper *twitterscraper.Scraper) ([]*TweetResult, error) {
+		var tweets []*TweetResult
+		ctx := context.Background()
+		scraper.SetSearchMode(twitterscraper.SearchLatest)
+		for tweet := range scraper.SearchTweets(ctx, query, count) {
+			if tweet.Error != nil {
+				return nil, tweet.Error
 			}
-			continue
+			tweets = append(tweets, &TweetResult{Tweet: &tweet.Tweet})
 		}
-		tweets = append(tweets, &TweetResult{Tweet: &tweetResult.Tweet, Error: nil})
+		return tweets, nil
 	}
 
-	if len(tweets) == 0 && lastError != nil {
-		return nil, lastError
+	handleRateLimit := func(err error, account *TwitterAccount) bool {
+		if strings.Contains(err.Error(), "Rate limit exceeded") {
+			accountManager.MarkAccountRateLimited(account)
+			logrus.Warnf("rate limited: %s", account.Username)
+			return true
+		}
+		return false
 	}
 
-	return tweets, nil
+	return retryTweets(func() ([]*TweetResult, error) {
+		scraper, account, err := getAuthenticatedScraper()
+		if err != nil {
+			return nil, err
+		}
+
+		tweets, err := scrapeTweets(scraper)
+		if err != nil {
+			if handleRateLimit(err, account) {
+				return nil, err
+			}
+			return nil, err
+		}
+		return tweets, nil
+	}, maxRetries)
 }
 
-// ScrapeTweetsByTrends scrapes the current trending topics on Twitter.
-// It returns a slice of strings representing the trending topics.
-// If an error occurs during the scraping process, it returns an error.
-func ScrapeTweetsByTrends() ([]*TweetResult, error) {
-	scraper := auth()
-	var trendResults []*TweetResult
+func ScrapeTweetsProfile(username string) (twitterscraper.Profile, error) {
+	once.Do(initializeAccountManager)
```

Comment:
why having a singleton here? can we pass the account manager, or have this method part of the account manager and pass it down the line to the other components? alternatively have another structure which depends on the account manager would sound a good option as well